includes/class-llms-events.php

```diff
@@ -0,0 +1,411 @@
+	 *
+	 * @since [version]
+	 *
+	 * @param array[] $events Array of event hashes. See LLMS_Events::record() for hash description.
+	 * @return LLMS_Event[]|WP_Error Array of recorded events on success or WP_Error on failure.
+	 */
+	public function record_many( $events = array() ) {
+
+		global $wpdb;
+		$wpdb->query( 'START TRANSACTION' );
+
+		$recorded = array();
+		$errors   = array();
+		foreach ( $events as $event ) {
+
+			$stat = $this->record( $event );
+			if ( is_wp_error( $stat ) ) {
+				$stat->add_data( $event );
+				$errors[] = $stat;
+			} else {
+				$recorded[] = $stat;
+			}
+		}
+
+		if ( count( $errors ) ) {
+			$wpdb->query( 'ROLLBACK' );
+			return new WP_Error( 'llms_events_record_many_errors', __( 'There was one or more errors encountered while recording the events.', 'lifterlms' ), $errors );
+		}
+
+		$wpdb->query( 'COMMIT' );
+
+		return $recorded;
+
+	}
+
+	/**
+	 * Recursively sanitize event data.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $raw Event information array.
+	 * @return array
+	 */
+	protected function sanitize_raw_event( $raw ) {
+
+		$clean = array();
+
+		foreach ( $raw as $key => $val ) {
+
+			// This will recursively handle any metadata submitted.
+			if ( is_array( $val ) ) {
+				$val = $this->sanitize_raw_event( $val );
+			} elseif ( in_array( $key, array( 'actor_id', 'object_id' ), true ) ) {
+				// cast id fields to int.
+				$val = absint( $val );
+			} else {
+				// everything else is a text field.
+				$val = sanitize_text_field( $val );
+			}
+
+			// Sanitize the key. This will ensure no dirty keys are submitted in metadata.
+			$key = is_numeric( $key ) ? $key : sanitize_text_field( $key );
+
+			$clean[ $key ] = $val;
+
+		}
+
+		return $clean;
+
+	}
+
+	/**
+	 * Determine if client side events from the current page should be tracked.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean
+	 */
+	protected function should_track_client_events() {
+
+		$ret = false;
+
+		/**
+		 * Filter the post types that should be tracked
+		 *
+		 * @since [version]
+		 *
+		 * @param string[]|string $post_types An array of post type names or a pre-defined setting as a string.
+		 *                                    "llms" uses all public LifterLMS and LifterLMS Add-on post types.
+		 *                                    "all" tracks everything.
+		 */
+		$post_types = apply_filters( 'llms_tracking_post_types', 'llms' );
+
+		if ( 'all' === $post_types ) {
+			$ret = true;
+		} elseif ( 'llms' === $post_types ) {
+
+			// Filter public post types to include LifterLMS public post types.
+			$post_types = array_keys( get_post_types( array( 'public' => true ) ) );
+			foreach ( $post_types as $key => $type ) {
+				if ( ! in_array( $type, array( 'course', 'lesson' ), true ) && 0 !== strpos( $type, 'llms_' ) ) {
+					unset( $post_types[ $key ] );
+				}
+			}
+		}
+
+		if ( ! is_array( $post_types ) ) {
+			$ret = false;
+		} elseif ( is_singular() && is_singular( $post_types ) ) {
```

Comment:
I seem to remember issues where archive pages would return `true` on `is_singular( $post_types )` when called on a custom post type archive. This, I think, was the result of the last item in the loop being evaluated on the underlying `$wp_query` -- however this might be one of those things that I identified as the solution to another issue. Can you test this without the "redundant" checks and see if any funny business like this happens? If we're safe drop them for sure. Thanks <3.

Reply:
fixed in #935